source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "envoy/common/time.h"
+#include "envoy/config/filter/http/adaptive_concurrency/v2alpha/adaptive_concurrency.pb.h"
+#include "envoy/http/filter.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/concurrency_controller.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrency {
+
+/**
+ * Configuration for the adaptive concurrency limit filter.
+ */
+class AdaptiveConcurrencyFilterConfig {
+public:
+  AdaptiveConcurrencyFilterConfig(
+      const envoy::config::filter::http::adaptive_concurrency::v2alpha::AdaptiveConcurrency&
+          adaptive_concurrency,
+      Runtime::Loader& runtime, std::string stats_prefix, Stats::Scope& scope,
+      TimeSource& time_source);
+
+  Runtime::Loader& runtime() { return runtime_; }
+  const std::string& statsPrefix() const { return stats_prefix_; }
+  Stats::Scope& scope() { return scope_; }
+  TimeSource& timeSource() { return time_source_; }
+
+private:
+  Runtime::Loader& runtime_;
+  const std::string stats_prefix_;
+  Stats::Scope& scope_;
+  TimeSource& time_source_;
+};
+
+using AdaptiveConcurrencyFilterConfigSharedPtr = std::shared_ptr<AdaptiveConcurrencyFilterConfig>;
+using ConcurrencyControllerSharedPtr =
+    std::shared_ptr<ConcurrencyController::ConcurrencyController>;
+
+/**
+ * A filter that samples request latencies and dynamically adjusts the request
+ * concurrency window.
+ */
+class AdaptiveConcurrencyFilter : public Http::StreamFilter, Logger::Loggable<Logger::Id::filter> {
```

Comment:
It's probably better to derive from pass_through_filter.h for this filter and just override the functions you need.